### PR TITLE
[PIR] Implement Translator for Deprecated Operator (mul op) in PIR

### PR DESCRIPTION
**Pull request by DrRyanHuang**
### PR types

Others
### PR changes

Others
### Description

Others
- https://github.com/PaddlePaddle/Paddle/issues/56521

### CODE REVIEW

**Comment by DrRyanHuang** [comment]
@kangguangli 非常不好意思，打扰您了

```
Op mul arg out should be optional if it can be empty
```

想问一下您，我是需要重写一下 `OpTranscriber::GenerateOperationOutput` 方法吗? 

把这一行注释掉嘛?
```c++
      IR_ENFORCE(info.optional,
                 "Op %s arg %s should be optional if it can be empty",
                 op_desc.Type(),
                 legacy_output_name);
```


```c++
2023-09-05 21:26:30 ======================================================================
2023-09-05 21:26:30 ERROR: test_check_output (test_mul_op.TestMulOp2)
2023-09-05 21:26:30 ----------------------------------------------------------------------
2023-09-05 21:26:30 Traceback (most recent call last):
2023-09-05 21:26:30   File "/workspace/Paddle/build/test/legacy_test/test_mul_op.py", line 90, in test_check_output
2023-09-05 21:26:30     self.check_output(check_dygraph=False)
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 2605, in check_output
2023-09-05 21:26:30     res = self.check_output_with_place(
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 2459, in check_output_with_place
2023-09-05 21:26:30     static_checker.check()
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 2117, in check
2023-09-05 21:26:30     self.calculate_output()
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 2125, in calculate_output
2023-09-05 21:26:30     outs, fetch_list = self.op_test._calc_output(
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 1513, in _calc_output
2023-09-05 21:26:30     self._check_ir_output(place, program, feed_map, fetch_list, outs)
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 1395, in _check_ir_output
2023-09-05 21:26:30     ir_outs = executor.run(
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 1635, in run
2023-09-05 21:26:30     res = self._run_impl(
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 1812, in _run_impl
2023-09-05 21:26:30     program, new_exe = self._executor_cache.get_program_and_executor(
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 890, in get_program_and_executor
2023-09-05 21:26:30     return self._get_cached_program_and_executor(
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 1011, in _get_program_and_executor
2023-09-05 21:26:30     new_exe = _StandaloneExecutor(place, plan, scope)
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 787, in __init__
2023-09-05 21:26:30     self._new_exe = self._create_new_executor()
2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 811, in _create_new_executor
2023-09-05 21:26:30     new_exe = core.StandaloneExecutor(self._place, self._plan, self._scope)
2023-09-05 21:26:30 RuntimeError: Error occured at: ../paddle/fluid/ir_adaptor/translator/op_translator.cc:434 :
2023-09-05 21:26:30 Op mul arg out should be optional if it can be empty
```

**Comment by kangguangli** [comment]
> @kangguangli 非常不好意思，打扰您了
> 
> ```
> Op mul arg out should be optional if it can be empty
> ```
> 
> 想问一下您，我是需要重写一下 `OpTranscriber::GenerateOperationOutput` 方法吗?
> 
> 把这一行注释掉嘛?
> 
> ```c++
>       IR_ENFORCE(info.optional,
>                  "Op %s arg %s should be optional if it can be empty",
>                  op_desc.Type(),
>                  legacy_output_name);
> ```
> 
> ```c++
> 2023-09-05 21:26:30 ======================================================================
> 2023-09-05 21:26:30 ERROR: test_check_output (test_mul_op.TestMulOp2)
> 2023-09-05 21:26:30 ----------------------------------------------------------------------
> 2023-09-05 21:26:30 Traceback (most recent call last):
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/test/legacy_test/test_mul_op.py", line 90, in test_check_output
> 2023-09-05 21:26:30     self.check_output(check_dygraph=False)
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 2605, in check_output
> 2023-09-05 21:26:30     res = self.check_output_with_place(
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 2459, in check_output_with_place
> 2023-09-05 21:26:30     static_checker.check()
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 2117, in check
> 2023-09-05 21:26:30     self.calculate_output()
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 2125, in calculate_output
> 2023-09-05 21:26:30     outs, fetch_list = self.op_test._calc_output(
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 1513, in _calc_output
> 2023-09-05 21:26:30     self._check_ir_output(place, program, feed_map, fetch_list, outs)
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/eager_op_test.py", line 1395, in _check_ir_output
> 2023-09-05 21:26:30     ir_outs = executor.run(
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 1635, in run
> 2023-09-05 21:26:30     res = self._run_impl(
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 1812, in _run_impl
> 2023-09-05 21:26:30     program, new_exe = self._executor_cache.get_program_and_executor(
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 890, in get_program_and_executor
> 2023-09-05 21:26:30     return self._get_cached_program_and_executor(
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 1011, in _get_program_and_executor
> 2023-09-05 21:26:30     new_exe = _StandaloneExecutor(place, plan, scope)
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 787, in __init__
> 2023-09-05 21:26:30     self._new_exe = self._create_new_executor()
> 2023-09-05 21:26:30   File "/workspace/Paddle/build/python/paddle/fluid/executor.py", line 811, in _create_new_executor
> 2023-09-05 21:26:30     new_exe = core.StandaloneExecutor(self._place, self._plan, self._scope)
> 2023-09-05 21:26:30 RuntimeError: Error occured at: ../paddle/fluid/ir_adaptor/translator/op_translator.cc:434 :
> 2023-09-05 21:26:30 Op mul arg out should be optional if it can be empty
> ```
@DrRyanHuang 
这种情况理论上需要添加参数映射，错误原因是PT模块安装OpYamlInfo从OpDesc中查找需要的信息，OpYamlInfo标记的输出是out，但是实际上OpDesc中存储的是Out，我们需要标记这种映射关系。

可以尝试修改`op_compat.yaml`，如下：
```
- op : matmul_with_flatten (mul)
  backward : matmul_with_flatten_grad (mul_grad) 
  inputs :
    {x : X, y : Y}
  outputs :
    out : Out
  extra :
    attrs : [bool use_mkldnn = false, float scale_x = 1.0f, 'float[] scale_y = {1.0f}',
             float scale_out = 1.0f, bool force_fp32_output = false]
```
这里写的可以参考下。我们对mul的处理比较特殊，如果这样后续有问题的话，可以再@下我。

**Comment by DrRyanHuang** [comment]
@kangguangli 不好意思，打扰了
```
Op mul should have corresponding OpInfo pd.matmul
```
应该还需要改yaml文件 `op_compat.yaml` 吧

将 `pd.matmul` 在 `op_compat.yaml` 中有吧：

```c
- op : matmul (matmul_v2)
  backward : matmul_grad (matmul_v2_grad), matmul_double_grad (matmul_v2_grad_grad), matmul_triple_grad (matmul_v2_triple_grad)
  inputs :
    {x : X, y : Y}
  attrs :
    {transpose_x : trans_x, transpose_y : trans_y}
  outputs :
    out : Out
  extra :
    attrs : [bool use_mkldnn = false, str mkldnn_data_type = "float32"]
  complex_promote : [X, Y]
```
为什么报没有找到呢?

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 1]
```diff
@@ -1240,6 +1240,267 @@ struct TrilAndTriuOpTranscriber : public OpTranscriber {
   }
 };
 
+struct MulOpTranscriber : public OpTranscriber {
+  pir::OpInfo LoopkUpOpInfo(pir::IrContext* ctx,
+                            const OpDesc& op_desc) override {
+    const std::string& target_op_name = "pd_op.matmul";
```
Comment (Aurelius84) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc:1252]: 后续这里建议可以优化为：
```suggestion
    const std::string& target_op_name = paddle::dialect::MatmulOp::name();
```

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 2]
```diff
@@ -1240,6 +1240,267 @@ struct TrilAndTriuOpTranscriber : public OpTranscriber {
...
+      pir::Block* block) override {
+    int x_num_col_dims = paddle::get<int>(op_desc.GetAttr("x_num_col_dims"));
+    int y_num_col_dims = paddle::get<int>(op_desc.GetAttr("y_num_col_dims"));
+
+    auto x_names = op_desc.Input("X", true);
+    IR_ENFORCE(x_names.size() == 1,
+               "Expected op[%s]'s input X has only 1 variable, but got %d",
+               op_desc.Type(),
+               x_names.size());
+    auto x_name = x_names[0];
```
Comment (Aurelius84) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc:1290]:
```suggestion
    auto& x_name = x_names[0];
```
可以了解下 auto 与 auto& 的区别

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 3]
```diff
@@ -1240,6 +1240,267 @@ struct TrilAndTriuOpTranscriber : public OpTranscriber {
...
+    IR_ENFORCE(param_map->count(x_name) > 0,
+               "Expected op[%s]'s input %s has been parsed",
+               op_desc.Type(),
+               x_name);
+    auto x_defining_info = param_map->at(x_name);
```
Comment (Aurelius84) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc:1295]:
```suggestion
    auto& x_defining_info = param_map->at(x_name);
```

Comment (Aurelius84) [reply paddle/fluid/ir_adaptor/translator/op_translator.cc:1295]: 可以系统性的review下使用 auto 的地方。另外对于pir中的Value 和 Result 数据结构可以不适用的auto& ,虽然也会触发copy构造，但因为是 Pimpl ，这个成本是极小的。

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 4]
```diff
@@ -1240,6 +1240,267 @@ struct TrilAndTriuOpTranscriber : public OpTranscriber {
...
+            << y_tensor_type.dims() << " to " << phi::make_ddim(y_new_shape);
+
+    return {x_new, y_new};
+  }
+};
+
+struct MulGradOpTranscriber : public OpTranscriber {
+  pir::OpInfo LoopkUpOpInfo(pir::IrContext* ctx,
+                            const OpDesc& op_desc) override {
+    const std::string& target_op_name = "pd_op.matmul_grad";
```
Comment (Aurelius84) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc:1527]:
```suggestion
    const std::string& target_op_name = paddle::dialect::MatmulGradOp::name();
```

**Review by Aurelius84 — DISMISSED** [review]
LGTM overall，遗留的comment可以单独提PR优化

File: test/white_list/new_ir_op_test_white_list [hunk 5]
```diff
@@ -214,3 +214,5 @@ test_warprnnt_op
 test_where_op
 test_yolo_box_op
 test_yolov3_loss_op
+test_fill_constant_op
+test_mul_op
```
Comment (kangguangli) [inline test/white_list/new_ir_op_test_white_list]: 提交前可以执行下以下命令：
```
cat test/white_list/new_ir_op_test_white_list | sort | uniq > test/white_list/new_ir_op_test_white_list
```
按字母序排下序，可以避免很多冲突

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 6]
```diff
@@ -1835,6 +2096,9 @@ OpTranslator::OpTranslator() {
   special_handlers["split"] = SplitOpTranscriber();
   special_handlers["sum"] = AddNOpTranscriber();
   special_handlers["tril_triu"] = TrilAndTriuOpTranscriber();
+  special_handlers["mul"] = MulOpTranscriber();
+  special_handlers["mul_grad"] = MulGradOpTranscriber();
+  special_handlers["fill_constant"] = FillConstantTranscriber();
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc]:
```suggestion

```
应该是解决冲突的时候搞错了，这句要删掉

**Comment by DrRyanHuang** [comment]
目前还有一个小bug:

image

33个test，25个跳过，7个通过，1个失败，应该还是反向 mul_grad 的问题

```
FLAGS_NEW_IR_OPTEST_WHITE_LIST=1 FLAGS_NEW_IR_OPTEST=1 GLOG_v=10 python test/legacy_test/test_mul_op.py  >err.log 2>&1
```

**Review by Aurelius84 — APPROVED** [review]
comment 里的建议下个PR 统一fix下~~

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 7]
```diff
@@ -1245,6 +1246,586 @@ struct TrilAndTriuOpTranscriber : public OpTranscriber {
...
+                 x_grad_var_name,
+                 x_grad_type);
+      dialect::DenseTensorType x_grad_tensor_type =
+          x_grad_type.dyn_cast<dialect::DenseTensorType>();
+
+      VLOG(10) << "[" << op_desc.Type() << "] x_grad_shape change from "
+               << x_grad_tensor_type.dims() << " to " << dim_x;
+
+      param_map->PushValue(x_grad_var_name,
+                           VariableDefiningInfo(reshape_op_x.out(), false, -1));
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc:1778]: 这段代码和下面基本一致，建议抽象为一个函数

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 8]
```diff
@@ -1245,6 +1246,586 @@ struct TrilAndTriuOpTranscriber : public OpTranscriber {
...
+    IR_ENFORCE(param_map->count(x_name) > 0,
+               "Expected op[%s]'s input %s has been parsed",
+               op_desc.Type(),
+               x_name);
+    auto x_defining_info = param_map->at(x_name);
+    if (x_defining_info.generated_by_vector) {
+      InsertSliceOperationForTarget(
+          ctx, param_map, block, x_defining_info, x_name);
+      x_defining_info = param_map->at(x_name);
+    }
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc:1300]: 这类代码应该不需要了，可以看看 `OpTranscriber::operator()`，里面会集中为输入插入slice op，一般不需要特殊处理。下面的类似处理也可以去掉

**Review by kangguangli — APPROVED** [review]
LGTM overall. But there're two problems that we should consider.
1. insert two much reshape, every pair of mul/mul_grad will lead to 6-10 additional reshape op insertions, is there some way to decrease this number?
2. too much redundant code. Try to compress by some kind of abstraction like function.

**Review by zyfncg — APPROVED** [review]
LGTM for op_compat

**Pull request merged**